src/realm/object-store/sync/async_open_task.cpp

```diff
@@ -41,9 +41,13 @@ void AsyncOpenTask::start(std::function<void(ThreadSafeReference, std::exception
     std::shared_ptr<AsyncOpenTask> self(shared_from_this());
     session->wait_for_download_completion([callback, self, this](std::error_code ec) {
         auto session = m_session.exchange(nullptr);
+
         if (!session)
-            return; // Swallow all events if the task as been canceled.
+            return; // Swallow all events if the task has been cancelled.
 
+        for (auto token : m_registered_callbacks) {
+            session->unregister_progress_notifier(token);
+        }
```

Comment:
There's a race condition between this and register_download_progress_notifier(). If the completion callback is called after `auto session = m_session.load()` but before `m_registered_callbacks.emplace_back(token);`, the mutation and iteration will race.

This probably just needs a mutex guarding it, which means that we can replace the `AtomicSharedPtr` with a normal `shared_ptr`. Note that it'll need to release the mutex before calling the callback.